File: frontend/lib/api/sync.ts
import { apiClient } from './client'

export interface SyncActionRequest {
  id: string
  type: string
  payload: any
  created_at: string
}

export interface SyncPushRequest {
  device_id?: string
  actions: SyncActionRequest[]
}

export interface SyncPushResponse {
  success: boolean
  processed_ids: string[]
  failed_ids: string[]
  errors: Record<string, string>
}

export interface SyncChange {
  type: string
  entity_id: number
  data: any
  updated_at: string
  action: string
}

export interface SyncPullResponse {
  server_time: string
  changes: SyncChange[]
  has_more: boolean
}

export interface SyncState {
  last_sync_at: string | null
  last_pull_at: string | null
  sync_version: number
}

export const syncApi = {
  // Push sync actions
  async push(data: SyncPushRequest): Promise<SyncPushResponse> {
    const response = await apiClient.post<SyncPushResponse>('/sync/push', data)
    return response.data
  },

  // Pull changes
  async pull(since?: string): Promise<SyncPullResponse> {
    const response = await apiClient.get<SyncPullResponse>('/sync/pull', {
      params: since ? { since } : {},
    })
    return response.data
  },

  // Get sync state
  async getState(): Promise<SyncState> {
    const response = await apiClient.get<SyncState>('/sync/state')
    return response.data
  },
}

File: backend/app/schemas/sync.py
"""
Sync schemas for request/response validation
"""
from pydantic import BaseModel
from typing import List, Dict, Any, Optional
from datetime import datetime


class SyncActionRequest(BaseModel):
    id: str  # Client-generated UUID
    type: str  # sale, stock_update, product_update, invoice
    payload: Dict[str, Any]
    created_at: datetime


class SyncPushRequest(BaseModel):
    device_id: Optional[str] = None
    actions: List[SyncActionRequest]


class SyncPushResponse(BaseModel):
    success: bool
    processed_ids: List[str]
    failed_ids: List[str]
    errors: Dict[str, str] = {}


class SyncChange(BaseModel):
    type: str  # product, stock, sale, invoice
    entity_id: int
    data: Dict[str, Any]
    updated_at: datetime
    action: str  # created, updated, deleted


class SyncPullResponse(BaseModel):
    server_time: datetime
    changes: List[SyncChange]
    has_more: bool = False

